File: PRACTICAS-PDOO/IRRGARTEN/Versiones_en_Java_Ruby/PROYECTO4-JAVA/src/irrgarten/Shield.java

/**
 * Esta clase representa los escudos que utiliza el jugador cuando se defiende 
 * de un ataque de un monstruo.
 * @file Shield.java
 * @author angel_rodriguez
 */

package irrgarten;

public class Shield extends CombatElement{

    /**
     * Constructor con parámetros
     * @param protection protección del escudo.
     * @param uses usos que le quedan al escudo.
     */
    public Shield (float protection, int uses){
        super(protection, uses);
    }
    
    /**
     * Constructor por defecto.
     */
    public Shield () {
        this(0f, 0);
    }
    
    /**
     * Devuelve un número representando la intensidad de defensa del jugador.
     * Sólo devolverá un número si uses > 0, si no, devuelve 0;
     * @return float representando la protección.
     */
    public float protect () {
        return super.produceEffect();
    }
    
    /**
     * Método toString.
     * @return representación del estado interno del arma. S[protection, uses]
     */
    @Override
    public String toString () {     
        return "Shield: " + super.produceEffect();
 
    }  
    
//    /**
//     * Implementa la decisión de si el escudo debe ser descartado.
//     * Delega en el método discardElement de la clase Dice.
//     * Hay que pasarle como parámetros la cantidad de usos restantes del escudo.
//     * @see Dice.discardElement(usesLeft).
//     * @return boolean si el escudo se descarta o no.
//     */
//    public boolean discard(){
//        return Dice.discardElement(uses);
//    }
    
}

File: PRACTICAS-PDOO/IRRGARTEN/Versiones_en_Java_Ruby/PROYECTO1-RUBY/shield.rb
# frozen_string_literal: true
#encoding:utf-8
require_relative 'dice'

module Irrgarten

  # Esta clase representa los escudos que utiliza el jugador cuando se defiende de un ataque de un monstruo.
  class Shield
    def initialize (protection, uses)
      @protection = protection.to_f
      @uses = uses.to_i
    end

    def protect
      if @uses > 0
        @uses -= 1
        return @protection
      else
        return 0.0
      end
    end

    def to_s
      "S[#{@protection}, #{@uses}]"
    end

    def discard
      Dice.discard_element(@uses)
    end

  end   # end de la clase Shield

end  # end del módulo Irrgarten